File: 662.Maximum-Width-of-Binary-Tree.py
# https://leetcode.com/problems/split-array-into-consecutive-subsequences/description/
#
# algorithms
# Medium (39.94%)
# Total Accepted:    26,874
# Total Submissions: 67,287
# beats 100.0% of python submissions


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque


class Solution(object):
    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        queue = deque([(root, 1)])
        res = 0

        while queue:
            l, r = queue[0][1], queue[-1][1]
            res = max(res, r - l + 1)
            for _ in xrange(len(queue)):
                cur_node, n = queue.popleft()
                if cur_node.left:
                    queue.append((cur_node.left, 2 * n))
                if cur_node.right:
                    queue.append((cur_node.right, 2 * n + 1))

        return res

File: 662.Maximum-Width-of-Binary-Tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }

        LinkedList<QueueItem> q = new LinkedList<>();
        q.offer(new QueueItem(0, root));
        q.offer(null);
        int res = Integer.MIN_VALUE;
        int l = -1, r = -1;

        while (true) {
            QueueItem item = q.pollFirst();
            if (item == null) {
                res = Math.max(res, r - l + 1);
                l = -1;
                r = -1;
                if (q.size() == 0) {
                    break;
                }
                q.offer(null);
            } else {
                if (l == -1) {
                    l = item.idx;
                }
                r = item.idx;
                if (item.node.left != null) {
                    q.offer(new QueueItem(item.idx * 2 + 1, item.node.left));
                }
                if (item.node.right != null) {
                    q.offer(new QueueItem(item.idx * 2 + 2, item.node.right));
                }
            }
        }

        return res;
    }

    class QueueItem {
        private int idx;
        private TreeNode node;

        QueueItem(int idx, TreeNode node) {
            this.idx = idx;
            this.node = node;
        }
    }
}